src/caustics/core.py

```diff
@@ -0,0 +1,43 @@
+import functools
+import inspect
+
+import torch
+
+
+def _args_to_kwargs(func, args):
+    signature = inspect.signature(func)
+    return {k: v for k, v in zip(signature.parameters.keys(), args)}
+
+
+def _sync_device(kwargs):
+    # Get all the tensor device types
+    tensor_device_types = {
+        k: v.device.type for k, v in kwargs.items() if isinstance(v, torch.Tensor)
+    }
+
+    # Get the set of devices
+    device_set = set([t for t in tensor_device_types.values() if t != "cpu"])
+    if device_set:
+        # Get the first device
+        device = device_set.pop()
+        new_tensors = {
+            k: kwargs[k].to(device)
+            for k, v in tensor_device_types.items()
+            if v == "cpu"
+        }
+        kwargs.update(new_tensors)
+    return kwargs
+
+
+def sync_device(func):
+    @functools.wraps(func)
+    def wrapper_decorator(*args, **kwargs):
+        kwargs.update(_args_to_kwargs(func, args))
+        kwargs = _sync_device(kwargs)
+
+        # Do something before
+        value = func(**kwargs)
+        # Do something after
+        return value
+
+    return wrapper_decorator
```

Comment:
Should this be in a new core functionality or utils? Need thoughts @uwcdc and @ConnorStoneAstro

Reply:
If I understand correctly, this is a wrapper to check that all tensors given as input to a function are on the same device. Is that right? I don't think that's necessary to include in caustics. Pytorch already throws an error if the tensors are not on the same device, which is the expected behavior. While it would be convenient to force the tensors onto the same device, it can result in hidden performance penalties if two functions end up "fighting" with eachother and successively pull a bunch of tensors between cpu and gpu. So it's better to throw the error and then have the user find the problem in their code for why they are passing tensors from different devices I think (@AlexandreAdam ?)